org-code-javabuilder/app/src/main/java/org/code/javabuilder/JavaRunner.java

```diff
@@ -0,0 +1,182 @@
+import javax.tools.StandardLocation;
+import javax.tools.ToolProvider;
+import org.springframework.beans.factory.config.ConfigurableBeanFactory;
+import org.springframework.context.annotation.Scope;
+import org.springframework.stereotype.Component;
+
+@Component
+@Scope(value = ConfigurableBeanFactory.SCOPE_SINGLETON)
+public class JavaRunner {
+
+  /**
+   * Compile and Run the given user program and output results to the given compileRunService for
+   * the given user (principal)
+   *
+   * @param userProgram
+   * @param principal
+   * @param compileRunService
+   */
+  public void compileAndRunUserProgram(
+      UserProgram userProgram, Principal principal, CompileRunService compileRunService) {
+
+    if (userProgram.getFileName().indexOf(".java") > 0) {
+      userProgram.setClassName(
+          userProgram.getFileName().substring(0, userProgram.getFileName().indexOf(".java")));
+    } else {
+      compileRunService.sendMessages(
+          principal.getName(), "Invalid File Name. File name must end in '.java'.");
+      return;
+    }
+
+    File tempFolder = null;
+    try {
+      tempFolder = Files.createTempDirectory("tmpdir").toFile();
+
+      compileRunService.sendMessages(principal.getName(), "Compiling your program...");
+      boolean compileSuccess =
+          compileProgram(userProgram, principal, compileRunService, tempFolder);
+
+      if (compileSuccess) {
+        // this try statement will close the streams automatically
+        try (ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+            PrintStream out = new PrintStream(outputStream)) {
+          // set System.out to be a specific output stream in order to capture output of the
+          // program and send it back to the user
+          System.setOut(out);
+
+          compileRunService.sendMessages(principal.getName(), "Compiled!");
+          compileRunService.sendMessages(principal.getName(), "Running your program...");
+          runClass(tempFolder.toURI().toURL(), userProgram, compileRunService, principal);
+
+          // outputStream should now contain output of userProgram
+          outputStream.flush();
+          String result = outputStream.toString();
+          if (result.length() > 0) {
+            compileRunService.sendMessages(principal.getName(), result);
+          }
+        }
+      } else {
+        compileRunService.sendMessages(
+            principal.getName(), "There was an error compiling your program.");
+      }
+
+    } catch (IOException e) {
+      // IOException could be called by creating a temporary folder or writing to that folder.
+      // May need better error handling for this.
+      compileRunService.sendMessages(
+          principal.getName(), "There was an issue trying to run your program, please try again.");
+      e.printStackTrace();
+    }
+
+    if (tempFolder != null) {
+      tempFolder.delete();
+    }
+
+    // ensure System.out is reset
+    System.setOut(System.out);
```

Comment:
I am catching all possible exceptions (the function itself doesn't throw any exceptions--shouldn't that mean I always get to the end of the function?

Reply:
The method can still throw unchecked exceptions.

IIRC, the whole checked vs. unchecked exceptions thing was really confusing in Java.  Some libraries (including Spring?) routinely throw unchecked exceptions because they viewed checked exceptions to be a language design flaw in Java.  Because of that, I think it is not generally safe to assume that all interesting exceptions are handled when your method signature does not have a throws clause.

Reply:
good context, thanks!